external-crates/move/crates/move-package/src/compilation/versioned_compilation.rs

```diff
@@ -0,0 +1,196 @@
+use move_symbol_pool::Symbol;
+
+use std::process::Command;
+use std::{
+    ffi::OsStr,
+    fs::File,
+    io::Write,
+    path::{Path, PathBuf},
+};
+use tracing::debug;
+
+/// partitions `deps` by whether we need to compile dependent packages with a
+/// prior toolchain (which we find by looking at Move.lock contents) or
+/// whether we can compile them with the current binary.
+pub fn partition_deps_by_toolchain<W: Write>(
+    deps: Vec<PackagePaths>,
+    current_compiler_version: Option<String>,
+    w: &mut W,
+) -> Result<(Vec<PackagePaths>, Vec<PackagePaths>)> {
+    let current_compiler_version = current_compiler_version.unwrap_or_else(|| "0.0.0".into());
+    debug!("current compiler: {current_compiler_version}");
+    let mut deps_for_current_compiler = vec![];
+    let mut deps_for_prior_compiler = vec![];
+    for dep in deps {
+        let a_source_path = dep.paths[0].as_str();
+        let root = SourcePackageLayout::try_find_root(Path::new(a_source_path))?;
+        let lock_file = root.join("Move.lock");
+        if !lock_file.exists() {
+            deps_for_current_compiler.push(dep);
+            continue;
+        }
+
+        let mut lock_file = File::open(lock_file)?;
+        let toolchain_version = lock_file::schema::ToolchainVersion::read(&mut lock_file)?;
+        match toolchain_version {
+            // No ToolchainVersion implies current compiler
+            None => deps_for_current_compiler.push(dep),
+            // This dependency uses the current compiler
+            Some(ToolchainVersion {
+                compiler_version, ..
+            }) if compiler_version == current_compiler_version => {
+                deps_for_current_compiler.push(dep)
+            }
+            // This dependency needs a prior compiler. Mark it and compile.
+            Some(toolchain_version) => {
+                let dep_name = match dep.name.clone() {
+                    Some((name, _)) => name,
+                    None => "unnamed dependency".into(),
+                };
+                writeln!(
+                    w,
+                    "{} {} compiler @ {}",
+                    "REQUIRE".bold().green(),
+                    dep_name,
+                    toolchain_version.compiler_version.yellow(),
+                )?;
+                download_and_compile(root, toolchain_version, dep_name, w)?;
+                deps_for_prior_compiler.push(dep)
+            }
+        }
+    }
+    Ok((deps_for_current_compiler, deps_for_prior_compiler))
+}
+
+fn detect_platform() -> Result<String> {
+    let s = match (std::env::consts::OS, std::env::consts::ARCH) {
+        ("macos", "aarch64") => "macos-arm64",
+        ("macos", "x86_64") => "macos-x86_64",
+        ("linux", "x86_64") => "ubuntu-x86_64",
+        ("windows", "x86_64") => "windows-x86_64",
+        (os, arch) => bail!("unsupported os {os} and arch {arch}"),
+    };
+    Ok(s.into())
+}
+
+fn download_and_compile<W: Write>(
```

Comment:
I would like this function to not use `curl`/`tar` to download dependencies, and will iterate on this to use Rust libraries to do this. Mainly so that we don't require users to have these binaries installed.

Right now though I'm focused on an end-to-end solution and using `curl`/`tar` makes it straightforward to test before using Rust native functions.